_posts/2024-11-19-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,623 @@
+
+        let res: BooleanArray = BooleanBuffer::from_iter(values).into();
+
+        res.into_data().to_pyarrow(py)
+    }
+}
+```
+
+We convert the `values_of_interest` into a vector of borrowed types so that we can do a fast search
+without creating additional memory. The other option is to turn the `returnflag` into a `String`
+but that memory allocation is unnecessary. After that we use two `zip` operations so that we can
+iterate over all three columns in a single pass. Since each `zip` will return a tuple of two
+elements, a quick `map` turns them into the tuple format we need. Also, `StringArray` is a little
+different in the buffer it uses, so it is treated slightly differently from the others.
+
+## User Defined Aggregate Function
+
+Writing a user defined aggregate function or user defined window function is slightly more complex
+than scalar functions. This is because we must accumulate values and there is no guarantee that one
+batch will contain all the values we are aggregating over. For this we need to define an
+`Accumulator` which will do a few things.
+
+- Process a batch and compute an internal state
+- Share the state so that we can combine multiple batches
+- Merge the results across multiple batches
+- Return the final result
+
+In the example below, we're going to look at customer orders and we want to know per customer ID,
+how much they have ordered total. We want to ignore small orders, which we define as anything under
+5000.
+
+```python
+from datafusion import Accumulator, udaf
+import pyarrow as pa
+import pyarrow.compute as pc
+
+IGNORE_THESHOLD = 5000.0
+class AboveThresholdAccum(Accumulator):
+    def __init__(self) -> None:
+        self._sum = 0.0
+
+    def update(self, values: pa.Array) -> None:
+        over_threshold = pc.greater(values, pa.scalar(IGNORE_THESHOLD))
+        sum_above = pc.sum(values.filter(over_threshold)).as_py()
+        if sum_above is None:
+            sum_above = 0.0
+        self._sum = self._sum + sum_above
+
+    def merge(self, states: List[pa.Array]) -> None:
+        self._sum = self._sum + pc.sum(states[0]).as_py()
+
+    def state(self) -> List[pa.Scalar]:
+        return [pa.scalar(self._sum)]
+
+    def evaluate(self) -> pa.Scalar:
+        return pa.scalar(self._sum)
+
+sum_above_threshold = udaf(AboveThresholdAccum, [pa.float64()], pa.float64(), [pa.float64()], 'stable')
+
+df_orders.aggregate([col("o_custkey")],[sum_above_threshold(col("o_totalprice")).alias("sales")]).show()
+```
+
+Since we are doing a `sum` we can keep a single value as our internal state. When we call `update()`
+we will process a single array and update the internal state, which we share with the `state()`
+function. For larger batches we may `merge()` these states. It is important to note that the
+`states` in the `merge()` function are an array of the values returned from `state()`. It is
+entirely possible that the `merge` function is significantly different than the `update`, though in
```

Comment:
A classic example is "avg" where the state is a sum + count and the output is a value

Reply:
`avg` was my first example before I found a bug (https://github.com/apache/datafusion-python/issues/797)  and switched it to sum to unblock myself!